File: python/exacto/exon.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""
The purpose of this python3 script is to implement the Exon dataclass.
"""


import pandas as pd
from dataclasses import dataclass, field
from functools import total_ordering
from typing import List
from .nucleotide_sequence import NucleotideSequence


@total_ordering
@dataclass(frozen=True)
class Exon:
    id: str                                         # e.g. 'ENSE00001910415.10'
    stable_id: str                                  # e.g. 'ENSE00001910415'
    source: str                                     # e.g. 'ENSEMBL'
    source_version: str                             # e.g. '107'
    chromosome: str                                 # e.g. 'chr1'
    start: int                                      # e.g. 1000 (5' start)
    end: int                                        # e.g. 1200 (3' end)
    number: int                                     # e.g. 1
    strand: str                                     # e.g. '+'
    sequence: NucleotideSequence
    utr_start: int = None                           # e.g. 10000
    utr_end: int = None                             # e.g. 10500
    version: int = None                             # e.g. 10
    tags: List[str] = field(default_factory=list)   # e.g. 'Ensembl_canonical'

    @property
    def length(self) -> int:
        return len(self.sequence.sequence)

    def __lt__(self, other):
        if isinstance(other, Exon):
            return self.number < other.number
        return NotImplemented

    def __eq__(self, other):
        if isinstance(other, Exon):
            return self.id == other.id
        return NotImplemented

    def to_dict(self):
        data = {
            'exon_id': [self.id],
            'exon_stable_id': [self.stable_id],
            'exon_source': [self.source],
            'exon_source_version': [self.source_version],
            'exon_chromosome': [self.chromosome],
            'exon_start': [self.start],
            'exon_end': [self.end],
            'exon_number': [self.number],
            'exon_strand': [self.strand],
            'exon_sequence': [self.sequence.sequence],
            'exon_utr_start': ['' if self.utr_start is None else self.utr_start],
            'exon_utr_end': ['' if self.utr_end is None else self.utr_end],
            'exon_version': [self.version],
            'exon_tags': [';'.join(self.tags)]
        }
        return data

    def to_dataframe(self):
        return pd.DataFrame(self.to_dict())

File: exacto/exacto-core/src/structs/gene_annotation/exon.rs
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


use serde::{Serialize, Deserialize};

use crate::common::constants::Strand;


#[derive(Debug, Serialize, Deserialize)]
pub struct Exon {
    pub gene_id: Box<str>,
    pub transcript_id: Box<str>,
    pub exon_id: Box<str>,
    pub source: Box<str>,
    pub chromosome: Box<str>,
    pub start: usize,
    pub end: usize,
    pub strand: Strand,
    pub level: u8,
    pub exon_number: u32
}

impl PartialEq for Exon {
    fn eq(&self, other: &Self) -> bool {
        if self.gene_id == other.gene_id &&
            self.transcript_id == other.transcript_id &&
            self.exon_id == other.exon_id &&
            self.source == other.source &&
            self.chromosome == other.chromosome &&
            self.start == other.start &&
            self.end == other.end &&
            self.strand == other.strand &&
            self.level == other.level &&
            self.exon_number == other.exon_number {
            true
        } else {
            false
        }
    }
}


impl Exon {
    pub fn new(
        gene_id: &str,
        transcript_id: &str,
        exon_id: &str,
        source: &str,
        chromosome: &str,
        start: usize,
        end: usize,
        strand: Strand,
        level: u8,
        exon_number: u32
    ) -> Self {
        Self {
            gene_id: gene_id.to_string().into_boxed_str(),
            transcript_id: transcript_id.to_string().into_boxed_str(),
            exon_id: exon_id.to_string().into_boxed_str(),
            source: source.to_string().into_boxed_str(),
            chromosome: chromosome.to_string().into_boxed_str(),
            start: start,
            end: end,
            strand: strand.clone(),
            level: level,
            exon_number: exon_number
        }
    }

    pub fn get_size(&self) -> usize {
        self.end - self.start + 1
    }
}

impl Clone for Exon {
    fn clone(&self) -> Self {
        Exon {
            gene_id: self.gene_id.to_string().into_boxed_str(),
            transcript_id: self.transcript_id.to_string().into_boxed_str(),
            exon_id: self.exon_id.to_string().into_boxed_str(),
            source: self.source.to_string().into_boxed_str(),
            chromosome: self.chromosome.to_string().into_boxed_str(),
            start: self.start,
            end: self.end,
            strand: self.strand.clone(),
            level: self.level,
            exon_number: self.exon_number
        }
    }
}
